### ISSUE TITLE
Use corfu-margin-formatters

### ISSUE THREAD

**Issue by minad** [jdtsmith/kind-icon#6]
Note that I removed the kind formatter with the more general margin formatters: https://github.com/minad/corfu/commit/61e520bafff48cc4d9bce30b17068c0a7c674630

See also https://github.com/minad/corfu/pull/75#issuecomment-966918321

**Comment by jdtsmith** [comment]
I thought we had discussed keeping both.  The kind-formatter interface is appealingly simple.  This new margin-formatter appears to require a function to go find the metadata itself, which requires unnecessary repetitive inquiry of the table function, and seems less reliable.  I'd prefer to stick to the kind formatter, though corfu can of course offer the more general formatter as an option, so plug-ins can pick whichever works for them.

**Comment by minad** [comment]
No point in keeping both. See https://github.com/minad/corfu/pull/75#issuecomment-966918321 for the adapter.

**Comment by minad** [comment]
> This new margin-formatter appears to require a function to go find the metadata itself, which requires unnecessary repetitive inquiry of the table function, and seems less reliable.

This is not the case. The margin function is curried. The metadata inquiry happens once, not repetitively.

**Comment by jdtsmith** [comment]
I don't see the `corfu-kind-formatter` helper you mentioned in that comment, which is the essence of what I meant by "keeping both". 

> This is not the case. The margin function is curried. The metadata inquiry happens once, not repetitively.

By this I mean corfu has already queried the metadata once, in `corfu--recompute-candidates`. 

I must admit I'm not understanding the motivation for this change.  It does add generality, but at the cost of decreasing the separation of concerns and adding unnecessary coupling between corfu and a hypothetic margin-function package.  It also adds complexity for "add-on" package authors.  

Unless corfu does the translating itself, any margin-function package enabled to work with it will be _required_ to consult corfu internals.  And if corfu does the translating itself, it hasn't gained anything.  All a margin-function-function has to work with is the candidate string.  I could see passing kind _and_ candidate, and letting it optionally decide whether to use or ignore the candidate name it was passed.  

Is there a specific use case for this generalization you envision that might help motivate it further?  I understand the "file suffix" case, but that could easily be done by simply passing the candidate as an optional 2nd argument to the kind-formatter.

**Comment by minad** [comment]
> I don't see the corfu-kind-formatter helper you mentioned in that comment, which is the essence of what I meant by "keeping both".

I haven't added it yet. I can add it, but you could also just add it to your package at the small cost of a little corfu dependency. The company margin functions also require company dependence, so that's at least not worse.

>  It does add generality, but at the cost of decreasing the separation of concerns and adding unnecessary coupling between corfu and a hypothetic margin-function package.

True. But the generality is worth it. As I argued company-kind is not sufficient for all use cases.

> It also adds complexity for "add-on" package authors.

Come on :) You added the affixation function advice to your package - then the three line wrapper should hardly be a problem. 
But of course introducing a corfu dependent piece of code is not pretty.

> I could see passing kind and candidate, and letting it optionally decide whether to use or ignore the candidate name it was passed.

This is not a good idea as I argued previously. The kind is not always available.  :company-kind is not always available. An icon library could decide to add icons based on the completion category for example.

> I understand the "file suffix" case

This is the prime example use case.

>  I understand the "file suffix" case, but that could easily be done by simply passing the candidate as an optional 2nd argument to the kind-formatter.

Incorrect. For files the kind is not available.

**Comment by minad** [comment]
I assume your package stays focused on kinds only? The icons you have won't fit that well it other contexts because of their background, e.g. file icons in this style in dired or in minibuffer completion.

Are you aware of @iyefrat's https://github.com/iyefrat/all-the-icons-completion? It uses fonts, not svgs and hooks into minibuffer completion similar to your package via the affixation-function.

Ideally a package could fit all the use cases. This is what I am adapting for and for this reason I went with the `corfu-margin-formatters` instead of the more limited `corfu-kind-formatter`.

**Comment by minad** [comment]
For reference:

~~~ elisp
(add-hook 'corfu-margin-formatters #'kind-example)
(add-hook 'corfu-margin-formatters #'file-example)

;; works in prog-mode buffers
(defun kind-example ()
  (when-let (kind (plist-get corfu--extra :company-kind))
    (lambda (cand)
      (format "<%s> " (funcall kind cand)))))

;; works in eshell
(defun file-example ()
  (when (eq (corfu--metadata-get 'category) 'file)
    (lambda (cand)
      (format "<%s> " (file-name-extension cand)))))
~~~

And one could even invent more margin functions depending on arbitrary metadata, which is made available by a completion table.

**Comment by jdtsmith** [comment]
> > It does add generality, but at the cost of decreasing the separation of concerns and adding unnecessary coupling between corfu and a hypothetic margin-function package.
> 
> True. But the generality is worth it. As I argued company-kind is not sufficient for all use cases.
> 
> > It also adds complexity for "add-on" package authors.
> 
> But of course introducing a corfu dependent piece of code is not pretty.

Yes, this. It's API has proved troublingly unstable ;) [1].   Kind formatter was something anybody could trivially implement a front-end for, even UI's that _don't yet exist_.  Margin formatter loses this.  Is such a new UI likely to come to pass?  Hard to say.  But why preclude it unnecessarily?

> > I could see passing kind and candidate, and letting it optionally decide whether to use or ignore the candidate name it was passed.
> 
> This is not a good idea as I argued previously. The kind is not always available. :company-kind is not always available. An icon library could decide to add icons based on the completion category for example.

If that's the case, why don't you hand the function the entire ball of metadata as a plist with all the "usual" slots, and let it decide if it wants to do anything with it?

> > I understand the "file suffix" case
> 
> This is the prime example use case.

Which could easily be solved without any hard-coded corfu dependency by passing candidate and kind (and if strictly necessary, which I doubt, the full ball of metadata).  A proper kind-less kind-formatter could even ignore the kind and just look at the candidate. 
 
> > I understand the "file suffix" case, but that could easily be done by simply passing the candidate as an optional 2nd argument to the kind-formatter.
> 
> Incorrect. For files the kind is not available.

I didn't say the formatter had to use the kind.  See above.  

**Concrete proposal**: retain the name margin-formatter, but instead of having it dig deep into (possibly changing) corfu internals, pass it candidate and a plist of metadata to (optionally) chew on.

[1] This is mild teasing with a nugget of truth.  It's early on, so obviously now's the time to hammer out the API, but please consider the need for agreement on both sides of the API pipe!

**Comment by jdtsmith** [comment]
> Are you aware of @iyefrat's https://github.com/iyefrat/all-the-icons-completion? It uses fonts, not svgs and hooks into minibuffer completion similar to your package via the affixation-function.

I've seen it but haven't looked closely.  Some of the fonts include the basic (google-created) "material" icons, but not all the ever-growing list of community add-ons, which is why svg-lib's approach was attractive.

**Comment by minad** [comment]
> If that's the case, why don't you hand the function the entire ball of metadata as a plist with all the "usual" slots, and let it decide if it wants to do anything with it?

This sounds like something worth to consider. However the metadata format is already quite baroque (completion-extra-properties and the metadata of the completion table), such that it wouldn't lead to a simplification. Furthermore you may want to use other criteria to decide which formatting function to use.

> Which could easily be solved without any hard-coded corfu dependency by passing candidate and kind.

I try to explain this better. Kind and company-kind are not always available. Corfu disables the default margin if a margin formatter is present. Therefore Corfu has to distinguish the scenarios: company-kind present/formatter present vs not present. One could change the protocol such that Corfu always passes kind (possibly nil if not available) and candidate to the kind formatter. Then if the kind formatter returns nil for some candidate the default margin is used. This will be more complicated and uglier on the side of Corfu. The idea is that the margin function can decide early on if it can provide icons. If it cannot, it returns nil. Does this make sense?

> This is mild teasing with a nugget of truth. It's early on, so obviously now's the time to hammer out the API, ...

The API has not been released and I was trying different approaches. So it is not fair to say that anything has been unstable so far. Furthermore any of these plugin APIs will not have the same degree of stability until more components have settled. The small interface we are discussing here is trivial in comparison to other API agreements we had to do for example across Consult and Embark, or also regarding how completion tables and capfs are treated. I don't expect difficulties even if a margin formatter plugin author has to dig a little bit into Corfu internals. This is over discussing and overthinking it - in particular if one contrasts it with the `affixation-function` code you have in your package. This code is severely less robust and more problematic, yet it is there.

**Comment by minad** [comment]
> Concrete proposal: retain the name margin-formatter, but instead of having it dig deep into (possibly changing) corfu internals, pass it candidate and a plist of metadata to (optionally) chew on.

Thinking about it - this idea is not that bad. With that the example would look like this:

~~~ elisp
(add-hook 'corfu-margin-formatters #'kind-example)
(add-hook 'corfu-margin-formatters #'file-example)

;; works in prog-mode buffers
(defun kind-example (_metadata)
  (when-let (kind (plist-get completion-extra-properties :company-kind))
    (lambda (cand)
      (format "<%s> " (funcall kind cand)))))

;; works in eshell
(defun file-example (metadata)
  (when (eq (completion-metadata-get metadata 'category) 'file)
    (lambda (cand)
      (format "<%s> " (file-name-extension cand)))))
~~~

There is no Corfu dependency anymore. However the special signature (and the currying) of the margin formatter still ties this quite closely to Corfu such that in practice not much is won. But I am still thinking about this. It is a clean API.

**Comment by jdtsmith** [comment]
> > Which could easily be solved without any hard-coded corfu dependency by passing candidate and kind.
> 
> I try to explain this better. Kind and company-kind are not always available. Corfu disables the default margin if a margin formatter is present. Therefore Corfu has to distinguish the scenarios: company-kind present/formatter present vs not present. One could change the protocol such that Corfu always passes kind (possibly nil if not available) and candidate to the kind formatter. Then if the kind formatter returns nil for some candidate the default margin is used. This will be more complicated and uglier on the side of Corfu. The idea is that the margin function can decide early on if it can provide icons. If it cannot, it returns nil. Does this make sense?

This does make sense, though you could as you say pass a `nil` kind (and my formatter would just return nil).
So you could configure a bunch of them in your margin-formatters list, and one of them could say "you've given me no kind and a candidate? I'll take a pass" and let another on the list have a try.  I can see the appeal there.  Since you're expecting the middle layer _margin formatters_ to do padding, you'll need to have some clear docs saying "if your formatter returns non-nil, I'll be disabling margin, so pad accordingly."  

> > This is mild teasing with a nugget of truth. It's early on, so obviously now's the time to hammer out the API, ...
> 
> The API has not been released and I was trying different approaches. 
 
I know, which is why the teasing was mild ;).  

> the `affixation-function` code you have in your package. This code is severely less robust and more problematic, yet it is there.

I agree, this is the disgusting "backup plan" for people who want to force icons into an uncooperative UI.  Hopefully we make it easy and simple for UI's to cooperate.  Kind-formatter was dead easy.  But not flexible enough for various envisioned use cases.  This new setup strikes me as falling too far on side of "super general purpose, but lots of overhead to do right." (kind of like CAPFs now that I think of it).  A happier middle is a good goal.

**Comment by jdtsmith** [comment]
> > Concrete proposal: retain the name margin-formatter, but instead of having it dig deep into (possibly changing) corfu internals, pass it candidate and a plist of metadata to (optionally) chew on.
> 
> Thinking about it - this idea is not that bad. With that the example would look like this:
> 
> ```emacs-lisp
> (add-hook 'corfu-margin-formatters #'kind-example)
> (add-hook 'corfu-margin-formatters #'file-example)
> 
> ;; works in prog-mode buffers
> (defun kind-example (_metadata)
>   (when-let (kind (plist-get completion-extra-properties :company-kind))
>     (lambda (cand)
>       (format "<%s> " (funcall kind cand)))))

How about:

```elisp
;; works in prog-mode buffers
(defun kind-example (kind _candidate &optional metadata)
  (let ((category (plist-get metadata :category)
	(some-prop-i-need (plist-get metadata :someprop)))
	(if (and (not (eq category 'some-useless-category))
		 some-prop-i-need)
	    (kind-formatted-icon kind some-prop-i-need)))))
```

**Comment by minad** [comment]
> This does make sense, though you could as you say pass a nil kind (and my formatter would just return nil).
So you could configure a bunch of them in your margin-formatters list, and one of them could say "you've given me no kind and a candidate? I'll take a pass" and let another on the list have a try. I can see the appeal there. 

No, this will be more complicated on the side of Corfu. I won't do this.

> Since you're expecting the middle layer margin formatters to do padding, you'll need to have some clear docs saying "if your formatter returns non-nil, I'll be disabling margin, so pad accordingly."

Sure, docs can be improved always.

> This new setup strikes me as falling too far on side of "super general purpose, but lots of overhead to do right." (kind of like CAPFs now that I think of it). A happier middle is a good goal.

I disagree that it is complicated. It is exactly as powerful as it should be :)

> How about: ...

No, I prefer the current design. The margin formatter should decide first if it is applicable. If yes, it returns the function which is then passed the candidates. In your proposal you single out `kind` for example - why is it so special? It is a company extension after all.

**Comment by minad** [comment]
I think I've made up my mind mostly about this and I want to keep the current API. The only thing which I consider debatable is if the metadata should be passed to the margin formatters as you proposed. See https://github.com/jdtsmith/kind-icon/issues/6#issuecomment-968327217.

The point is - the margin formatter API is exactly the generalization which was lacking in kind-formatter. The decision if a formatter is present is pushed to the plugin, which is a good idea, since as I argued at length kind is not all that is interesting.

**Comment by jdtsmith** [comment]
> you single out kind for example - why is it so special? It is a company extension after all. 

OK, do you at least agree _candidate_ is interesting!?!  So:

```elisp
(defun kind-example (candidate &optional metadata)
  (if-let ((kind-func (plist-get metadata :company-kind))
	   (kind (funcall kind-func candidate))
	   (some-prop-i-need (plist-get metadata :someprop)))
      (if (not (eq (plist-get metadata :category)
		   'some-useless-category))
	  (kind-formatted-icon kind some-prop-i-need))))
```

> The margin formatter should decide first if it is applicable. 

Yes, except this is not like the situation of CAPF's.  The expensive metadata have already been gathered.  So a formatter can make the decision of whether it will be able to return a result, and then return that result, all in one efficient and simple pass.

**Comment by minad** [comment]
> OK, do you at least agree candidate is interesting!?!

I think you misunderstood how the margin formatter functions work?

**Comment by jdtsmith** [comment]
> > OK, do you at least agree candidate is interesting!?!
> 
> I think you misunderstood how the margin formatter functions work?

I think I'm clear on how they should work: like kind-formatters, but based on arbitrary metadata and/or candidate text, instead of just on "kind".

**Comment by minad** [comment]
> I think I'm clear on how they should work: like kind-formatters, but based on arbitrary metadata and/or candidate text, instead of just on "kind".

Unfortunately this won't work. See https://github.com/jdtsmith/kind-icon/issues/6#issuecomment-968326011.

>  So a formatter can make the decision of whether it will be able to return a result, and then return that result, all in one efficient and simple pass.

The misunderstanding is this: While the single pass can of course work easily from the side of the plugin author it does not work well for Corfu, since in Corfu I have to decide ONCE if margins are to be enabled or not. So if the margin formatter is called multiple times and some calls returned nil and some calls returned a string, what should Corfu do? This ambiguous scenario is exactly what is disallowed by my API.

**Comment by jdtsmith** [comment]
I'm sure it doesn't really need pointing out, but this is due to the choice of making padding the domain of plugins, which I suspected would bite us sooner or later.  Apparently later has arrived...

I'll leave this alone for a bit and revisit later.

**Comment by minad** [comment]
> I'm sure it doesn't really need pointing out, but this is due to the choice of making padding the domain of plugins, which I suspected would bite us sooner or later. Apparently later has arrived...

There is no problem here. See it like this - the margin is either generated by Corfu (a thin margin at the left side of the popup with the background color) or the margin is populated by the margin formatter. The margin formatters belong to the frontend, as much as Corfu itself. So there is nothing wrong with it. We don't violate levels of responsibility (backend, presentation, etc).

**Comment by jdtsmith** [comment]
OK I've looked closer, and, I remain unconvinced.  The design doesn't appear to gain anything useful over a list of functions, each of which receives candidate+metadata, and each of which could return a string, with the first non-`nil` value "winning".  And in fact it even loses some flexibility which the user may sorely miss (see example below).  

In terms of the "single pass" issue, in my simpler suggested setup, upon accumulating a list in `corfu--affixate`, corfu can easily decide _once_ (per affixation) whether any prefixes are present, and remove the margin accordingly.  No prefixes?  Do not remove margin.  At least one non-nil prefix?  Remove.  

Unlike CAPF, which is the parallel of this design, a hypothetical margin-formatter-function has _no real context with which to decide_ if it should return a function to return prefixes.  In a CAPF, you can determine if you are in a string, comment, function body, etc.  Here you have only the metadata of the completion as potential context [1].  Note that _even with metadata in hand_, you still don't know what  flavor of candidates the function you dutifully provide will be presented with, so it's not really possible to make an informed decision whether to return to corfu a function.  And the metadata alone aren't helpful.  Recall that lsp-servers sometimes, randomly, _decide not to provide kind information in a timely manner_.  So even, as a margin-formatters-function, you _think_ you can "commit" to providing a formatted margin, you may be unable to do so during that particular completion! 

Let's see if I can make this even more concrete.  Suppose you are a `file-icon-margin-formatter`.  _All you do is files_.  You are very good at it.  You have icons for GIFs, JPG's, Excel files, Hypercard stack files, Amiga shared binary files, etc.  You've even hand-made some icons of your own.  You are more than willing to provide a formatted margin for each and every kind of file or directory ever in existence.  But nothing else.  When corfu comes asking, you do not know whether you will be able to provide margins for all the potential candidates, _since it hasn't yet asked for them_.   Suppose you say "yes" by returning a function.  While holding your function, corfu asks your function for "mytaxes.xls" and "kittens.jpg".  So far so good.  But then it asks for a `kind: 'variable`.  Danger, does not compute... return `nil`!  Meanwhile, the user had configured _another_ margin formatter, in 2nd position on their list: `super-kind-command-margin-formatter`.  That formatter specializes in LSP kinds.  But it never got a chance...

[1]  Side issue: not very conveniently, since the margin-formatter-function isn't given the metadata in the current setup.

**Comment by minad** [comment]
> In terms of the "single pass" issue, in my simpler suggested setup, upon accumulating a list in corfu--affixate, corfu can easily decide once (per affixation) whether any prefixes are present, and remove the margin accordingly. No prefixes? Do not remove margin. At least one non-nil prefix? Remove.

This is a decision step I don't want to perform in Corfu. My goal is to have a design where no flexibility is lost for both sides and where the amount of baggage in Corfu is minimized. My current design achieves this.

> Unlike CAPF, which is the parallel of this design, a hypothetical margin-formatter-function has no real context with which to decide if it should return a function to return prefixes.

Of course there is a context. It is the buffer and the buffer local variables, including for example the major mode and including the metadata which is hold by Corfu.

>  Recall that lsp-servers sometimes, randomly, decide not to provide kind information in a timely manner. So even, as a margin-formatters-function, you think you can "commit" to providing a formatted margin, you may be unable to do so during that particular completion!

I looked at both emacs lsp clients and both provide a company-kind function. Based on the existence of this property a margin formatter can take over. If then for some reason no kind is provided the margin formatter can print a place holder icon or do something else which yields a consistent look with the other candidates.

>  Meanwhile, the user had configured another margin formatter, in 2nd position on their list: super-kind-command-margin-formatter. That formatter specializes in LSP kinds. But it never got a chance...

The use case you describe here (mixing of different formatters) is intentionally precluded since it is a recipe for incoherent looks within a single completion session. I don't want to have the modest file icons mixed with the super kind icons which possibly come from different packages. Of course the margin formatter is still free to mix different icon sets, so technically no flexibility is lost. The intention of supporting multiple margin formatters in Corfu is to register different formatters for different major modes for example or for different completion categories (file, lsp-capf, ...). Note that the completion candidates only have a single completion category, no mixing of different candidate sources takes place. The intention is that for a single completion session only a single formatter function is ever used.

**Comment by jdtsmith** [comment]
Well, I guess we'll have to agree to disagree.  Do you intend to pass the margin-formatter-function-function the metadata list, in addition to the candidate?  Since corfu already has that in hand, it seems quite reasonable.  For extra completion properties, I suppose I can just consult `completion-extra-properties`, just as I do for the affixation hack method.

**Comment by minad** [comment]
I think will rather add the `corfu-kind-formatter` adapter mentioned in https://github.com/minad/corfu/pull/75#issuecomment-96691832 instead of passing metadata. It seems better to access arbitrary buffer local variables directly - as I said the margin formatter could also check the major mode for example and other data.

The adapter is used like this:

~~~ elisp
(setq corfu-margin-formatters (corfu-kind-formatter #'kind-icon-formatted))
~~~

For libraries interested in kinds only like yours this is sufficient. If other libraries have other requirements they can write their own margin formatters which provide icons based on other contextual information.

Is this an acceptable solution from your perspective?

**Comment by minad** [comment]
I added both in https://github.com/minad/corfu/commit/b16143d5a7affacace05a1097ba7d8680f38dea5

**Comment by minad** [comment]
Please ping me when you release your library. I will then tag a new Corfu version.

**Comment by jdtsmith** [comment]
Implemented in b88e809.  Did not mention the use of the wrapper function, if you'd prefer to omit.

**Comment by minad** [comment]
Okay, I dropped it. I will tag a new release soon!

**Comment by minad** [comment]
Do you plan to submit your package soon to ELPA or MELPA? And maybe make a post somewhere, e.g., on reddit, to give this a bit of visibility?

**Comment by jdtsmith** [comment]
I submitted it to emacs-devel for inclusion in ELPA last night; will post a brief announcement after it's up.  I've never submitted a package to ELPA;  anything to keep in mind?

**Comment by minad** [comment]
> I've never submitted a package to ELPA; anything to keep in mind?

I have no particular tips, but I am sure that Philip Kaludercic (@phikal) and Stefan Monnier will be helpful and supportive. I will tag a new Corfu release now, to make sure that your package works right away as soon as it appears in the ELPA archive.

**Comment by minad** [comment]
Released 0.15 https://github.com/minad/corfu/commit/3b01af4baa76bed05ec37f141ca8bc69f713c0c5

Not yet on ELPA, but I expect it there soon.
